### PR TITLE
[IR][NFC] Update IRBuilder to use InsertPosition

### PR DESCRIPTION
**Pull request by SLTozer**
Uses the new InsertPosition class (added in #94226) to simplify some of the IRBuilder interface, and removes the need to pass a BasicBlock alongside a BasicBlock::iterator, using the fact that we can now get the parent basic block from the iterator even if it points to the sentinel. This patch removes the BasicBlock argument from each constructor or call to setInsertPoint.

This has no functional effect, but later on as we look to remove the `Instruction *InsertBefore` argument from instruction-creation (discussed [here](https://discourse.llvm.org/t/psa-instruction-constructors-changing-to-iterator-only-insertion/77845)), this will simplify the process by allowing us to deprecate the InsertPosition constructor directly and catch all the cases where we use instructions rather than iterators.

### CODE REVIEW

File: llvm/include/llvm/IR/Instruction.h [hunk 1]
```diff
@@ -44,15 +44,19 @@ template <> struct ilist_alloc_traits<Instruction> {
 iterator_range<simple_ilist<DbgRecord>::iterator>
 getDbgRecordRange(DbgMarker *);
 
+/// Class used to generate an insert position (ultimately always a
+/// BasicBlock::iterator, which it will implicitly convert to) from either:
+/// - An Instruction, inserting immediately prior.
```
Comment (nikic) [inline llvm/include/llvm/IR/Instruction.h:49]: Indicate that this is deprecated?

File: llvm/lib/IR/Core.cpp [hunk 2]
```diff
@@ -3136,7 +3136,7 @@ static void LLVMPositionBuilderImpl(IRBuilder<> *Builder, BasicBlock *Block,
                                     Instruction *Instr, bool BeforeDbgRecords) {
   BasicBlock::iterator I = Instr ? Instr->getIterator() : Block->end();
   I.setHeadBit(BeforeDbgRecords);
-  Builder->SetInsertPoint(Block, I);
+  Builder->SetInsertPoint(I);
```
Comment (nikic) [inline llvm/lib/IR/Core.cpp:3139]: Maybe assert that BB and Instr are consistent here?

File: llvm/lib/Target/ARM/ARMParallelDSP.cpp [hunk 3]
```diff
@@ -637,8 +637,7 @@ void ARMParallelDSP::InsertParallelMACs(Reduction &R) {
         Intrinsic::getDeclaration(M, Intrinsic::arm_smlad) :
         Intrinsic::getDeclaration(M, Intrinsic::arm_smlald);
 
-    IRBuilder<NoFolder> Builder(InsertAfter->getParent(),
-                                BasicBlock::iterator(InsertAfter));
+    IRBuilder<NoFolder> Builder((BasicBlock::iterator(InsertAfter)));
```
Comment (nikic) [inline llvm/lib/Target/ARM/ARMParallelDSP.cpp:640]:
```suggestion
    IRBuilder<NoFolder> Builder(BasicBlock::iterator(InsertAfter));
```

**Review by nikic — APPROVED** [review]
LGTM if CI is happy.

File: llvm/lib/Target/ARM/ARMParallelDSP.cpp (hunk 3 above)
Comment (SLTozer) [reply llvm/lib/Target/ARM/ARMParallelDSP.cpp:640]: Funnily enough, that doesn't actually work:
```
./llvm/lib/Target/ARM/ARMParallelDSP.cpp:640:32: warning: parentheses were disambiguated as a function declaration [-Wvexing-parse]
  640 |     IRBuilder<NoFolder> Builder(BasicBlock::iterator(InsertAfter));
      |                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Hence the weird double parentheses - however having looked back at this I can see `InsertAfter` is always non-null, so `->getIterator()` is safe.

File: llvm/include/llvm/IR/Instruction.h (hunk 1 above)
Comment (SLTozer) [reply llvm/include/llvm/IR/Instruction.h:49]: It isn't deprecated yet, but I can put a comment indicating the future intent.

**Comment by SLTozer** [comment]
Looks like I missed a spot - reverting for now, will wait for a bit to see if any other errors fall out from the various configurations on CI before reapplying.

**Comment by SLTozer** [comment]
> Looks like using InsertPosition in IRBuilder has some overhead: 

Hm, we're adding an extra pointer chase when we give up passing a basicblock - I think then it makes sense to keep the option to pass BB+It, and do so when the caller already knows the BB (while removing the need to pass the BB around for functions that only need an insert point).

**Comment by nikic** [comment]
> > Looks like using InsertPosition in IRBuilder has some overhead:
> 
> Hm, we're adding an extra pointer chase when we give up passing a basicblock - I think then it makes sense to keep the option to pass BB+It, and do so when the caller already knows the BB (while removing the need to pass the BB around for functions that only have an instruction).

My guess here would have been that the issue is the cases where an Instruction* or BasicBlock* is passed rather than BB+Iterator, because those two cases go through an out-of-line constructor.

**Comment by SLTozer** [comment]
> My guess here would have been that the issue is the cases where an Instruction* or BasicBlock* is passed rather than BB+Iterator, because those two cases go through an out-of-line constructor.

Seems like you're right - removing the extra dereference does little-to-nothing, but allowing the constructors to be inlined mitigates most of the cost. It still ends up being [a net negative for performance](http://llvm-compile-time-tracker.com/compare.php?from=d75f9dd1d29b332bdc51346de63cbc04646354d7&to=3acef100b23fb6fa616523dd0d1b85945adb7900&stat=instructions%3Au) though, even after allowing the constructors to become inlined.

**Pull request merged**